### PR TITLE
Fix CSE detection in algebraic loops

### LINKED ISSUE #8164: Make sure $cse records computed by functions are always selected as torn variables.
### Description

Sometimes a few variables belonging to records that are computed by function calls are chosen as tearing variables in nonlinear systems. This is typically not a good tearing, and additionally those variables lack a proper nominal attribute, leading to potential issues as already noticed in #7800.

### Steps to Reproduce

Run `ThermoSysPro.Examples.SimpleExamples.TestStaticCondenser.mos`

This situation is obtained:
```
Variables (12)
========================================
1: $cse17.d:VARIABLE(min = max(1e-09, 0.0) max = 100000.0 start = 900.0 nominal = 998.0 protected = true )  type: Real unreplaceable
2: $cse17.duph:VARIABLE(protected = true )  type: Real unreplaceable
3: $cse17.x:VARIABLE(protected = true )  type: Real unreplaceable
4: $cse17.ddph:VARIABLE(protected = true )  type: Real unreplaceable
5: $cse17.s:VARIABLE(min = -1000000.0 max = 1000000.0 nominal = 1000.0 protected = true )  type: Real unreplaceable
6: $cse17.duhp:VARIABLE(protected = true )  type: Real unreplaceable
7: $cse17.ddhp:VARIABLE(protected = true )  type: Real unreplaceable
8: $cse17.u:VARIABLE(min = -100000000.0 max = 100000000.0 nominal = 1000000.0 protected = true )  type: Real unreplaceable
9: condenseur.Pex:VARIABLE(min = max(0.0, max(0.0, 0.0)) start = 100000.0 unit = "Pa" nominal = 100000.0 )  "Drain pressure at the outlet" type: Real
10: $cse17.cp:VARIABLE(min = 1e-09 max = 9.999999999999999e+59 nominal = 1000.0 protected = true )  type: Real unreplaceable
11: $cse17.T:VARIABLE(min = 200.0 max = 6000.0 start = 288.15 nominal = 320.0 protected = true )  type: Real unreplaceable
12: condenseur.Hex:VARIABLE(start = 500000.0 unit = "J/kg" )  "Drain specific enthalpy at the outlet" type: Real

Equations (3, 12)
========================================
1/1 (1): condenseur.Hex = if noEvent($cse17.d > 0.0) then condenseur.Hsate + (condenseur.Pex - condenseur.Pcond) / $cse17.d else condenseur.Hsate   [dynamic |0|0|0|0|]
2/2 (10): $cse17 = ThermoSysPro.Properties.WaterSteam.IF97.Water_Ph(condenseur.Pex, condenseur.Hex, condenseur.mode_ex)   [unknown |0|0|0|0|]
3/12 (1): condenseur.Pex = condenseur.Pcond + 9.806649999999999 * $cse17.d * condenseur.z   [dynamic |0|0|0|0|]
```

### Expected Behavior

`$cse17 = Water_Ph(...)` should be selected as a torn equation, so that the whole record `$cse17` is a torn variable.

Consequently, `condenseur.Hex` and `condenseur.Pex` are very likely to be chosen as tearing variables, which would be a much better choice.

### PR DESCRIPTION
**Pull request by phannebohm**
### Related Issues

Fixes #8164

Co-authored-by: kabdelhak <karim.abdelhak@fh-bielefeld.de>

### CODE REVIEW

File: testsuite/openmodelica/dataReconciliation/TSP_Pipe3.mos [hunk 1]
```diff
@@ -2247,15 +2247,24 @@ getErrorString();
...
 // stdout            | info    | DataReconciliation Starting!
 // stdout            | info    | NewDataReconciliationSimpleTests.TSP_Pipe3
+// stdout            | warning | The default linear solver fails, the fallback solver with total pivoting is started at time 1.000000. That might raise performance issues, for more information use -lv LOG_LS.
+// LOG_NLS           | error   | residualFunc76: Iteration variable xloc[0] is nan.
+// assert            | debug   | residualFunc76 failed at time=1.
+// |                 | |       | For more information please use -lv LOG_NLS.
+// assert            | debug   | Solving non-linear system 76 failed at time=1.
+// |                 | |       | For more information please use -lv LOG_NLS.
+// getBestJumpBuffer got simulationJumpBuffer=(nil)
+// Aborted (core dumped)
```
Comment (AnHeuermann) [inline testsuite/openmodelica/dataReconciliation/TSP_Pipe3.mos:2267]: 👀

Comment (phannebohm) [reply testsuite/openmodelica/dataReconciliation/TSP_Pipe3.mos:2267]: I'm not sure what's supposed to happen. The model (before dataReconciliation) runs now, previously it failed at initialization. @arun3688 any idea what's going on here?

Comment (phannebohm) [reply testsuite/openmodelica/dataReconciliation/TSP_Pipe3.mos:2267]: I'll merge this for now to see the regressions. We can fix this in another PR.

Comment (phannebohm) [reply testsuite/openmodelica/dataReconciliation/TSP_Pipe3.mos:2267]: See #10109 for further discussion.

**Pull request merged**